File: app/src/main/java/com/jbrunton/mymovies/app/search/SearchResultsAdapter.java
package com.jbrunton.mymovies.app.search;

import android.content.Context;
import android.content.Intent;
import android.support.v7.widget.RecyclerView;
import android.text.Html;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.jbrunton.mymovies.app.shared.BaseRecyclerAdapter;
import com.jbrunton.mymovies.R;
import com.jbrunton.mymovies.app.moviedetails.MovieDetailsActivity;
import com.squareup.picasso.Picasso;

public class SearchResultsAdapter extends BaseRecyclerAdapter<MovieViewState, SearchResultsAdapter.ViewHolder> {
    public SearchResultsAdapter(Context context, int layoutId) {
        super(layoutId, new ViewHolderFactory(context));
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        private TextView titleView;
        private TextView releaseDateView;
        private ImageView poster;
        private TextView ratingView;
        private View movieCardView;

        ViewHolder(View itemView) {
            super(itemView);
            movieCardView = itemView.findViewById(R.id.movie_card_view);
            titleView = (TextView) itemView.findViewById(R.id.title);
            releaseDateView = (TextView) itemView.findViewById(R.id.release_date);
            ratingView = (TextView) itemView.findViewById(R.id.rating);
            poster = (ImageView) itemView.findViewById(R.id.poster);
        }
    }

    protected static class ViewHolderFactory implements BaseRecyclerAdapter.ViewHolderFactory<MovieViewState, ViewHolder> {
        private final Context context;

        public ViewHolderFactory(Context context) {
            this.context = context;
        }

        @Override
        public ViewHolder createViewHolder(View view) {
            return new ViewHolder(view);
        }

        @Override
        public void bindHolder(ViewHolder holder, MovieViewState item) {
            holder.titleView.setText(item.title());
            if (holder.releaseDateView != null) {
                holder.releaseDateView.setText(item.yearReleased());
            }
            if (holder.ratingView != null) {
                holder.ratingView.setText(Html.fromHtml(item.rating()));
            }
            Picasso.with(context)
                    .load(item.posterUrl())
                    .resize(185, 275)
                    .centerCrop()
                    .into(holder.poster);
            holder.movieCardView.setOnClickListener(new View.OnClickListener() {
                @Override public void onClick(View view) {
                    Intent intent = new Intent(context, MovieDetailsActivity.class);
                    intent.putExtra("MOVIE_ID", item.movieId());
                    context.startActivity(intent);
                }
            });
        }
    }
}

File: app/src/main/java/com/jbrunton/mymovies/search/SearchResultsAdapter.kt
package com.jbrunton.mymovies.search

import android.content.Context
import android.content.Intent
import androidx.recyclerview.widget.RecyclerView
import android.text.Html
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import com.jbrunton.mymovies.R
import com.jbrunton.mymovies.moviedetails.MovieDetailsActivity
import com.jbrunton.mymovies.movies.MovieSearchResultViewState
import com.jbrunton.mymovies.shared.BaseRecyclerAdapter
import com.squareup.picasso.Picasso

class SearchResultsAdapter(
        context: Context,
        layoutId: Int
) : BaseRecyclerAdapter<MovieSearchResultViewState, SearchResultsAdapter.ViewHolder>(
        layoutId,
        ViewHolderFactory(context)
) {

    class ViewHolder(itemView: View) : androidx.recyclerview.widget.RecyclerView.ViewHolder(itemView) {
        val titleView: TextView
        val releaseDateView: TextView?
        val poster: ImageView
        val ratingView: TextView?
        val movieCardView: View

        init {
            movieCardView = itemView.findViewById(R.id.movie_card_view)
            titleView = itemView.findViewById<View>(R.id.title) as TextView
            releaseDateView = itemView.findViewById<View>(R.id.release_date) as TextView?
            ratingView = itemView.findViewById<View>(R.id.rating) as TextView?
            poster = itemView.findViewById<View>(R.id.poster) as ImageView
        }
    }

    class ViewHolderFactory(private val context: Context) : BaseRecyclerAdapter.ViewHolderFactory<MovieSearchResultViewState, ViewHolder> {

        override fun createViewHolder(view: View): ViewHolder {
            return ViewHolder(view)
        }

        override fun bindHolder(holder: ViewHolder, item: MovieSearchResultViewState) {
            holder.titleView.text = item.title
            if (holder.releaseDateView != null) {
                holder.releaseDateView.text = item.yearReleased
            }
            if (holder.ratingView != null) {
                holder.ratingView.text = Html.fromHtml(item.rating)
            }
            Picasso.with(context)
                    .load(item.posterUrl)
                    .resize(185, 275)
                    .centerCrop()
                    .into(holder.poster)
            holder.movieCardView.setOnClickListener {
                val intent = Intent(context, MovieDetailsActivity::class.java)
                intent.putExtra("MOVIE_ID", item.movieId)
                context.startActivity(intent)
            }
        }
    }
}
